# if (!requireNamespace("BiocManager", quietly = TRUE)) install.packages("BiocManager")
#BiocManager::install("StructuralVariantAnnotation")
#install.packages("stringr")
library(VariantAnnotation)
library(StructuralVariantAnnotation)
library(stringr)
#' Simple SV type classifier
simpleEventType <- function(gr) {
  pgr = partner(gr)
  return(ifelse(seqnames(gr) != seqnames(pgr), "CTX", # inter-chromosomosal
    ifelse(strand(gr) == strand(pgr), "INV",
      ifelse(gr$insLen >= abs(gr$svLen) * 0.7, "INS", # TODO: improve classification of complex events
        ifelse(xor(start(gr) < start(pgr), strand(gr) == "-"), "DEL",
          "DUP")))))
}

# Loading the GRIDSS output VCFs.
setwd("/media/jana/Data/R_Analysis/GRIDSS")

#####################################################################################################################

#M_vcf1 <- readVcf("vcf_male/males_B01.vcf.gz", "/media/jana/Data/kakpo125/genome/kakapo_full_ref.fa")
#info(header(M_vcf1)) = unique(as(rbind(as.data.frame(info(header(M_vcf1))), data.frame(
#	row.names=c("SIMPLE_TYPE"),
#	Number=c("1"),
#	Type=c("String"),
#	Description=c("Simple event type annotation based purely on breakend position and orientation."))), "DataFrame"))

#M_gr1 <- breakpointRanges(M_vcf1)
#svtype <- simpleEventType(M_gr1)
#info(M_vcf1)$SIMPLE_TYPE <- NA_character_
#info(M_vcf1[M_gr1$sourceId])$SIMPLE_TYPE <- svtype
#info(M_vcf1[M_gr1$sourceId])$SVLEN <- M_gr1$svLen
#writeVcf(M_vcf1, "GRIDSS_males_B01.sv.annotated.vcf") # generated by example/gridss.sh

# TODO: perform event filtering here
# By default, GRIDSS is very sensitive but this comes at the cost of a high false discovery rate
#M_gr1 <- M_gr1[M_gr1$FILTER == "PASS" & partner(M_gr1)$FILTER == "PASS"] # Remove low confidence calls
#simple_M_gr1 <- M_gr1[simpleEventType(M_gr1) %in% c("INS", "INV", "DEL", "DUP")]
#simplebed <- data.frame(
#	chrom=seqnames(simple_M_gr1),
# call the centre of the homology/inexact interval
#	start=as.integer((start(simple_M_gr1) + end(simple_M_gr1)) / 2),
#	end=as.integer((start(partner(simple_M_gr1)) + end(partner(simple_M_gr1))) / 2),
#	name=simpleEventType(simple_M_gr1),
#	score=simple_M_gr1$QUAL,
#	strand="."
#)

# Just the lower of the two breakends so we don't output everything twice
#simplebed <- simplebed[simplebed$start < simplebed$end,]
#write.table(simplebed, "GRIDSS_males_B01.simple.bed", quote=FALSE, sep='\t', row.names=FALSE, col.names=FALSE)

#####################################################################################################################

#M_vcf2 <- readVcf("vcf_male/males_B02.vcf.gz", "/media/jana/Data/kakpo125/genome/kakapo_full_ref.fa")
#info(header(M_vcf2)) = unique(as(rbind(as.data.frame(info(header(M_vcf2))), data.frame(
#  row.names=c("SIMPLE_TYPE"),
#  Number=c("1"),
#  Type=c("String"),
#  Description=c("Simple event type annotation based purely on breakend position and orientation."))), "DataFrame"))

#M_gr2 <- breakpointRanges(M_vcf2)
#M_svtype2 <- simpleEventType(M_gr2)
#info(M_vcf2)$SIMPLE_TYPE <- NA_character_
#info(M_vcf2[M_gr2$sourceId])$SIMPLE_TYPE <- M_svtype2
#info(M_vcf2[M_gr2$sourceId])$SVLEN <- M_gr2$svLen
#writeVcf(M_vcf2, "GRIDSS_males_B02.sv.annotated.vcf")

#M_gr2 <- M_gr2[M_gr2$FILTER == "PASS" & partner(M_gr2)$FILTER == "PASS"]
#simple_M_gr2 <- M_gr2[simpleEventType(M_gr2) %in% c("INS", "INV", "DEL", "DUP")]
#M_simplebed2 <- data.frame(
#  chrom=seqnames(simple_M_gr2),
#  start=as.integer((start(simple_M_gr2) + end(simple_M_gr2)) / 2),
#  end=as.integer((start(partner(simple_M_gr2)) + end(partner(simple_M_gr2))) / 2),
#  name=simpleEventType(simple_M_gr2),
#  score=simple_M_gr2$QUAL,
#  strand="."
#)

#M_simplebed2 <- M_simplebed2[M_simplebed2$start < M_simplebed2$end,]
#write.table(M_simplebed2, "GRIDSS_males_B02.simple.bed", quote=FALSE, sep='\t', row.names=FALSE, col.names=FALSE)

#####################################################################################################################

M_vcf3 <- readVcf("vcf_male/males_B03.vcf.gz", "/media/jana/Data/kakpo125/genome/kakapo_full_ref.fa")
info(header(M_vcf3)) = unique(as(rbind(as.data.frame(info(header(M_vcf3))), data.frame(
  row.names=c("SIMPLE_TYPE"),
  Number=c("1"),
  Type=c("String"),
  Description=c("Simple event type annotation based purely on breakend position and orientation."))), "DataFrame"))

M_gr3 <- breakpointRanges(M_vcf3)
M_svtype3 <- simpleEventType(M_gr3)
info(M_vcf3)$SIMPLE_TYPE <- NA_character_
info(M_vcf3[M_gr3$sourceId])$SIMPLE_TYPE <- M_svtype3
info(M_vcf3[M_gr3$sourceId])$SVLEN <- M_gr3$svLen
writeVcf(M_vcf3, "GRIDSS_males_B03.sv.annotated.vcf")

M_gr3 <- M_gr3[M_gr3$FILTER == "PASS" & partner(M_gr3)$FILTER == "PASS"]
simple_M_gr3 <- M_gr3[simpleEventType(M_gr3) %in% c("INS", "INV", "DEL", "DUP")]
M_simplebed3 <- data.frame(
  chrom=seqnames(simple_M_gr3),
  start=as.integer((start(simple_M_gr3) + end(simple_M_gr3)) / 2),
  end=as.integer((start(partner(simple_M_gr3)) + end(partner(simple_M_gr3))) / 2),
  name=simpleEventType(simple_M_gr3),
  score=simple_M_gr3$QUAL,
  strand="."
)

M_simplebed3 <- M_simplebed3[M_simplebed3$start < M_simplebed3$end,]
write.table(M_simplebed3, "GRIDSS_males_B03.simple.bed", quote=FALSE, sep='\t', row.names=FALSE, col.names=FALSE)

#####################################################################################################################

M_vcf4 <- readVcf("vcf_male/males_B04.vcf.gz", "/media/jana/Data/kakpo125/genome/kakapo_full_ref.fa")
info(header(M_vcf4)) = unique(as(rbind(as.data.frame(info(header(M_vcf4))), data.frame(
  row.names=c("SIMPLE_TYPE"),
  Number=c("1"),
  Type=c("String"),
  Description=c("Simple event type annotation based purely on breakend position and orientation."))), "DataFrame"))

M_gr4 <- breakpointRanges(M_vcf4)
M_svtype4 <- simpleEventType(M_gr4)
info(M_vcf4)$SIMPLE_TYPE <- NA_character_
info(M_vcf4[M_gr4$sourceId])$SIMPLE_TYPE <- M_svtype4
info(M_vcf4[M_gr4$sourceId])$SVLEN <- M_gr4$svLen
writeVcf(M_vcf4, "GRIDSS_males_B04.sv.annotated.vcf")

M_gr4 <- M_gr4[M_gr4$FILTER == "PASS" & partner(M_gr4)$FILTER == "PASS"]
simple_M_gr4 <- M_gr4[simpleEventType(M_gr4) %in% c("INS", "INV", "DEL", "DUP")]
M_simplebed4 <- data.frame(
  chrom=seqnames(simple_M_gr4),
  start=as.integer((start(simple_M_gr4) + end(simple_M_gr4)) / 2),
  end=as.integer((start(partner(simple_M_gr4)) + end(partner(simple_M_gr4))) / 2),
  name=simpleEventType(simple_M_gr4),
  score=simple_M_gr4$QUAL,
  strand="."
)

M_simplebed4 <- M_simplebed4[M_simplebed4$start < M_simplebed4$end,]
write.table(M_simplebed4, "GRIDSS_males_B04.simple.bed", quote=FALSE, sep='\t', row.names=FALSE, col.names=FALSE)

#####################################################################################################################

#M_vcf5 <- readVcf("vcf_male/males_B05.vcf.gz", "/media/jana/Data/kakpo125/genome/kakapo_full_ref.fa")
#info(header(M_vcf5)) = unique(as(rbind(as.data.frame(info(header(M_vcf5))), data.frame(
#  row.names=c("SIMPLE_TYPE"),
#  Number=c("1"),
#  Type=c("String"),
#  Description=c("Simple event type annotation based purely on breakend position and orientation."))), "DataFrame"))

#M_gr5 <- breakpointRanges(M_vcf5)
#M_svtype5 <- simpleEventType(M_gr5)
#info(M_vcf5)$SIMPLE_TYPE <- NA_character_
#info(M_vcf5[M_gr5$sourceId])$SIMPLE_TYPE <- M_svtype5
#info(M_vcf5[M_gr5$sourceId])$SVLEN <- M_gr5$svLen
#writeVcf(M_vcf5, "GRIDSS_males_B05.sv.annotated.vcf")

#M_gr5 <- M_gr5[M_gr5$FILTER == "PASS" & partner(M_gr5)$FILTER == "PASS"]
#simple_M_gr5 <- M_gr5[simpleEventType(M_gr5) %in% c("INS", "INV", "DEL", "DUP")]
#M_simplebed5 <- data.frame(
#  chrom=seqnames(simple_M_gr5),
#  start=as.integer((start(simple_M_gr5) + end(simple_M_gr5)) / 2),
#  end=as.integer((start(partner(simple_M_gr5)) + end(partner(simple_M_gr5))) / 2),
#  name=simpleEventType(simple_M_gr5),
#  score=simple_M_gr5$QUAL,
#  strand="."
#)

#M_simplebed5 <- M_simplebed5[M_simplebed5$start < M_simplebed5$end,]
#write.table(M_simplebed5, "GRIDSS_males_B05.simple.bed", quote=FALSE, sep='\t', row.names=FALSE, col.names=FALSE)

#####################################################################################################################

M_vcf6 <- readVcf("vcf_male/males_B06.vcf.gz", "/media/jana/Data/kakpo125/genome/kakapo_full_ref.fa")
info(header(M_vcf6)) = unique(as(rbind(as.data.frame(info(header(M_vcf6))), data.frame(
  row.names=c("SIMPLE_TYPE"),
  Number=c("1"),
  Type=c("String"),
  Description=c("Simple event type annotation based purely on breakend position and orientation."))), "DataFrame"))

M_gr6 <- breakpointRanges(M_vcf6)
M_svtype6 <- simpleEventType(M_gr6)
info(M_vcf6)$SIMPLE_TYPE <- NA_character_
info(M_vcf6[M_gr6$sourceId])$SIMPLE_TYPE <- M_svtype6
info(M_vcf6[M_gr6$sourceId])$SVLEN <- M_gr6$svLen
writeVcf(M_vcf6, "GRIDSS_males_B06.sv.annotated.vcf")

M_gr6 <- M_gr6[M_gr6$FILTER == "PASS" & partner(M_gr6)$FILTER == "PASS"]
simple_M_gr6 <- M_gr6[simpleEventType(M_gr6) %in% c("INS", "INV", "DEL", "DUP")]
M_simplebed6 <- data.frame(
  chrom=seqnames(simple_M_gr6),
  start=as.integer((start(simple_M_gr6) + end(simple_M_gr6)) / 2),
  end=as.integer((start(partner(simple_M_gr6)) + end(partner(simple_M_gr6))) / 2),
  name=simpleEventType(simple_M_gr6),
  score=simple_M_gr6$QUAL,
  strand="."
)

M_simplebed6 <- M_simplebed6[M_simplebed6$start < M_simplebed6$end,]
write.table(M_simplebed6, "GRIDSS_males_B06.simple.bed", quote=FALSE, sep='\t', row.names=FALSE, col.names=FALSE)

#####################################################################################################################

M_vcf7 <- readVcf("vcf_male/males_B07.vcf.gz", "/media/jana/Data/kakpo125/genome/kakapo_full_ref.fa")
info(header(M_vcf7)) = unique(as(rbind(as.data.frame(info(header(M_vcf7))), data.frame(
  row.names=c("SIMPLE_TYPE"),
  Number=c("1"),
  Type=c("String"),
  Description=c("Simple event type annotation based purely on breakend position and orientation."))), "DataFrame"))

M_gr7 <- breakpointRanges(M_vcf7)
M_svtype7 <- simpleEventType(M_gr7)
info(M_vcf7)$SIMPLE_TYPE <- NA_character_
info(M_vcf7[M_gr7$sourceId])$SIMPLE_TYPE <- M_svtype7
info(M_vcf7[M_gr7$sourceId])$SVLEN <- M_gr7$svLen
writeVcf(M_vcf7, "GRIDSS_males_B07.sv.annotated.vcf")

M_gr7 <- M_gr7[M_gr7$FILTER == "PASS" & partner(M_gr7)$FILTER == "PASS"]
simple_M_gr7 <- M_gr7[simpleEventType(M_gr7) %in% c("INS", "INV", "DEL", "DUP")]
M_simplebed7 <- data.frame(
  chrom=seqnames(simple_M_gr7),
  start=as.integer((start(simple_M_gr7) + end(simple_M_gr7)) / 2),
  end=as.integer((start(partner(simple_M_gr7)) + end(partner(simple_M_gr7))) / 2),
  name=simpleEventType(simple_M_gr7),
  score=simple_M_gr7$QUAL,
  strand="."
)

M_simplebed7 <- M_simplebed7[M_simplebed7$start < M_simplebed7$end,]
write.table(M_simplebed7, "GRIDSS_males_B07.simple.bed", quote=FALSE, sep='\t', row.names=FALSE, col.names=FALSE)

#####################################################################################################################

M_vcf8 <- readVcf("vcf_male/males_B08.vcf.gz", "/media/jana/Data/kakpo125/genome/kakapo_full_ref.fa")
info(header(M_vcf8)) = unique(as(rbind(as.data.frame(info(header(M_vcf8))), data.frame(
  row.names=c("SIMPLE_TYPE"),
  Number=c("1"),
  Type=c("String"),
  Description=c("Simple event type annotation based purely on breakend position and orientation."))), "DataFrame"))

M_gr8 <- breakpointRanges(M_vcf8)
M_svtype8 <- simpleEventType(M_gr8)
info(M_vcf8)$SIMPLE_TYPE <- NA_character_
info(M_vcf8[M_gr8$sourceId])$SIMPLE_TYPE <- M_svtype8
info(M_vcf8[M_gr8$sourceId])$SVLEN <- M_gr8$svLen
writeVcf(M_vcf8, "GRIDSS_males_B08.sv.annotated.vcf")

M_gr8 <- M_gr8[M_gr8$FILTER == "PASS" & partner(M_gr8)$FILTER == "PASS"]
simple_M_gr8 <- M_gr8[simpleEventType(M_gr8) %in% c("INS", "INV", "DEL", "DUP")]
M_simplebed8 <- data.frame(
  chrom=seqnames(simple_M_gr8),
  start=as.integer((start(simple_M_gr8) + end(simple_M_gr8)) / 2),
  end=as.integer((start(partner(simple_M_gr8)) + end(partner(simple_M_gr8))) / 2),
  name=simpleEventType(simple_M_gr8),
  score=simple_M_gr8$QUAL,
  strand="."
)

M_simplebed8 <- M_simplebed8[M_simplebed8$start < M_simplebed8$end,]
write.table(M_simplebed8, "GRIDSS_males_B08.simple.bed", quote=FALSE, sep='\t', row.names=FALSE, col.names=FALSE)
#####################################################################################################################

M_vcf9 <- readVcf("vcf_male/males_B09.vcf.gz", "/media/jana/Data/kakpo125/genome/kakapo_full_ref.fa")
info(header(M_vcf9)) = unique(as(rbind(as.data.frame(info(header(M_vcf9))), data.frame(
  row.names=c("SIMPLE_TYPE"),
  Number=c("1"),
  Type=c("String"),
  Description=c("Simple event type annotation based purely on breakend position and orientation."))), "DataFrame"))

M_gr9 <- breakpointRanges(M_vcf9)
M_svtype9 <- simpleEventType(M_gr9)
info(M_vcf9)$SIMPLE_TYPE <- NA_character_
info(M_vcf9[M_gr9$sourceId])$SIMPLE_TYPE <- M_svtype9
info(M_vcf9[M_gr9$sourceId])$SVLEN <- M_gr9$svLen
writeVcf(M_vcf9, "GRIDSS_males_B09.sv.annotated.vcf")

M_gr9 <- M_gr9[M_gr9$FILTER == "PASS" & partner(M_gr9)$FILTER == "PASS"]
simple_M_gr9 <- M_gr9[simpleEventType(M_gr9) %in% c("INS", "INV", "DEL", "DUP")]
M_simplebed9 <- data.frame(
  chrom=seqnames(simple_M_gr9),
  start=as.integer((start(simple_M_gr9) + end(simple_M_gr9)) / 2),
  end=as.integer((start(partner(simple_M_gr9)) + end(partner(simple_M_gr9))) / 2),
  name=simpleEventType(simple_M_gr9),
  score=simple_M_gr9$QUAL,
  strand="."
)

M_simplebed9 <- M_simplebed9[M_simplebed9$start < M_simplebed9$end,]
write.table(M_simplebed9, "GRIDSS_males_B09.simple.bed", quote=FALSE, sep='\t', row.names=FALSE, col.names=FALSE)