# if (!requireNamespace("BiocManager", quietly = TRUE)) install.packages("BiocManager")
#BiocManager::install("StructuralVariantAnnotation")
#install.packages("stringr")
library(VariantAnnotation)
library(StructuralVariantAnnotation)
library(stringr)
#' Simple SV type classifier
simpleEventType <- function(gr) {
  pgr = partner(gr)
  return(ifelse(seqnames(gr) != seqnames(pgr), "CTX", # inter-chromosomosal
    ifelse(strand(gr) == strand(pgr), "INV",
      ifelse(gr$insLen >= abs(gr$svLen) * 0.7, "INS", # TODO: improve classification of complex events
        ifelse(xor(start(gr) < start(pgr), strand(gr) == "-"), "DEL",
          "DUP")))))
}

# Loading the GRIDSS output VCFs.
setwd("/media/jana/Data/R_Analysis/GRIDSS")

#####################################################################################################################

F_vcf1 <- readVcf("vcf_female/Females_B01.vcf.gz", "/media/jana/Data/kakpo125/genome/kakapo_full_ref.fa")
info(header(F_vcf1)) = unique(as(rbind(as.data.frame(info(header(F_vcf1))), data.frame(
	row.names=c("SIMPLE_TYPE"),
	Number=c("1"),
	Type=c("String"),
	Description=c("Simple event type annotation based purely on breakend position and orientation."))), "DataFrame"))

F_gr1 <- breakpointRanges(F_vcf1)
svtype <- simpleEventType(F_gr1)
info(F_vcf1)$SIMPLE_TYPE <- NA_character_
info(F_vcf1[F_gr1$sourceId])$SIMPLE_TYPE <- svtype
info(F_vcf1[F_gr1$sourceId])$SVLEN <- F_gr1$svLen
writeVcf(F_vcf1, "GRIDSS_females_B01.sv.annotated.vcf") # generated by example/gridss.sh

# TODO: perform event filtering here
# By default, GRIDSS is very sensitive but this comes at the cost of a high false discovery rate
F_gr1 <- F_gr1[F_gr1$FILTER == "PASS" & partner(F_gr1)$FILTER == "PASS"] # Remove low confidence calls
simple_F_gr1 <- F_gr1[simpleEventType(F_gr1) %in% c("INS", "INV", "DEL", "DUP")]
F_simplebed1 <- data.frame(
	chrom=seqnames(simple_F_gr1), # call the centre of the homology/inexact interval
	start=as.integer((start(simple_F_gr1) + end(simple_F_gr1)) / 2),
	end=as.integer((start(partner(simple_F_gr1)) + end(partner(simple_F_gr1))) / 2),
	name=simpleEventType(simple_F_gr1),
	score=simple_F_gr1$QUAL,
	strand="."
)

# Just the lower of the two breakends so we don't output everything twice
F_simplebed1 <- F_simplebed1[F_simplebed1$start < F_simplebed1$end,]
write.table(F_simplebed1, "GRIDSS_females_B01.simple.bed", quote=FALSE, sep='\t', row.names=FALSE, col.names=FALSE)

#####################################################################################################################

F_vcf2 <- readVcf("vcf_female/Females_B02.vcf.gz", "/media/jana/Data/kakpo125/genome/kakapo_full_ref.fa")
info(header(F_vcf2)) = unique(as(rbind(as.data.frame(info(header(F_vcf2))), data.frame(
  row.names=c("SIMPLE_TYPE"),
  Number=c("1"),
  Type=c("String"),
  Description=c("Simple event type annotation based purely on breakend position and orientation."))), "DataFrame"))
F_gr2 <- breakpointRanges(F_vcf2)
F_svtype2 <- simpleEventType(F_gr2)
info(F_vcf2)$SIMPLE_TYPE <- NA_character_
info(F_vcf2[F_gr2$sourceId])$SIMPLE_TYPE <- F_svtype2
info(F_vcf2[F_gr2$sourceId])$SVLEN <- F_gr2$svLen
writeVcf(F_vcf2, "GRIDSS_females_B02.sv.annotated.vcf")

F_gr2 <- F_gr2[F_gr2$FILTER == "PASS" & partner(F_gr2)$FILTER == "PASS"]
simple_F_gr2 <- F_gr2[simpleEventType(F_gr2) %in% c("INS", "INV", "DEL", "DUP")]
F_simplebed2 <- data.frame(
  chrom=seqnames(simple_F_gr2),
  start=as.integer((start(simple_F_gr2) + end(simple_F_gr2)) / 2),
  end=as.integer((start(partner(simple_F_gr2)) + end(partner(simple_F_gr2))) / 2),
  name=simpleEventType(simple_F_gr2),
  score=simple_F_gr2$QUAL,
  strand="."
)

F_simplebed2 <- F_simplebed2[F_simplebed2$start < F_simplebed2$end,]
write.table(F_simplebed2, "GRIDSS_females_B02.simple.bed", quote=FALSE, sep='\t', row.names=FALSE, col.names=FALSE)

#####################################################################################################################

F_vcf3 <- readVcf("vcf_female/Females_B03.vcf.gz", "/media/jana/Data/kakpo125/genome/kakapo_full_ref.fa")
info(header(F_vcf3)) = unique(as(rbind(as.data.frame(info(header(F_vcf3))), data.frame(
  row.names=c("SIMPLE_TYPE"),
  Number=c("1"),
  Type=c("String"),
  Description=c("Simple event type annotation based purely on breakend position and orientation."))), "DataFrame"))

F_gr3 <- breakpointRanges(F_vcf3)
F_svtype3 <- simpleEventType(F_gr3)
info(F_vcf3)$SIMPLE_TYPE <- NA_character_
info(F_vcf3[F_gr3$sourceId])$SIMPLE_TYPE <- F_svtype3
info(F_vcf3[F_gr3$sourceId])$SVLEN <- F_gr3$svLen
writeVcf(F_vcf3, "GRIDSS_females_B03.sv.annotated.vcf")

F_gr3 <- F_gr3[F_gr3$FILTER == "PASS" & partner(F_gr3)$FILTER == "PASS"]
simple_F_gr3 <- F_gr3[simpleEventType(F_gr3) %in% c("INS", "INV", "DEL", "DUP")]
F_simplebed3 <- data.frame(
  chrom=seqnames(simple_F_gr3),
  start=as.integer((start(simple_F_gr3) + end(simple_F_gr3)) / 2),
  end=as.integer((start(partner(simple_F_gr3)) + end(partner(simple_F_gr3))) / 2),
  name=simpleEventType(simple_F_gr3),
  score=simple_F_gr3$QUAL,
  strand="."
)

F_simplebed3 <- F_simplebed3[F_simplebed3$start < F_simplebed3$end,]
write.table(F_simplebed3, "GRIDSS_females_B03.simple.bed", quote=FALSE, sep='\t', row.names=FALSE, col.names=FALSE)

#####################################################################################################################

F_vcf4 <- readVcf("vcf_female/Females_B04.vcf.gz", "/media/jana/Data/kakpo125/genome/kakapo_full_ref.fa")
info(header(F_vcf4)) = unique(as(rbind(as.data.frame(info(header(F_vcf4))), data.frame(
  row.names=c("SIMPLE_TYPE"),
  Number=c("1"),
  Type=c("String"),
  Description=c("Simple event type annotation based purely on breakend position and orientation."))), "DataFrame"))

F_gr4 <- breakpointRanges(F_vcf4)
F_svtype4 <- simpleEventType(F_gr4)
info(F_vcf4)$SIMPLE_TYPE <- NA_character_
info(F_vcf4[F_gr4$sourceId])$SIMPLE_TYPE <- F_svtype4
info(F_vcf4[F_gr4$sourceId])$SVLEN <- F_gr4$svLen
writeVcf(F_vcf4, "GRIDSS_females_B04.sv.annotated.vcf")

F_gr4 <- F_gr4[F_gr4$FILTER == "PASS" & partner(F_gr4)$FILTER == "PASS"]
simple_F_gr4 <- F_gr4[simpleEventType(F_gr4) %in% c("INS", "INV", "DEL", "DUP")]
F_simplebed4 <- data.frame(
  chrom=seqnames(simple_F_gr4),
  start=as.integer((start(simple_F_gr4) + end(simple_F_gr4)) / 2),
  end=as.integer((start(partner(simple_F_gr4)) + end(partner(simple_F_gr4))) / 2),
  name=simpleEventType(simple_F_gr4),
  score=simple_F_gr4$QUAL,
  strand="."
)

F_simplebed4 <- F_simplebed4[F_simplebed4$start < F_simplebed4$end,]
write.table(F_simplebed4, "GRIDSS_females_B04.simple.bed", quote=FALSE, sep='\t', row.names=FALSE, col.names=FALSE)

#####################################################################################################################

F_vcf5 <- readVcf("vcf_female/Females_B05.vcf.gz", "/media/jana/Data/kakpo125/genome/kakapo_full_ref.fa")
info(header(F_vcf5)) = unique(as(rbind(as.data.frame(info(header(F_vcf5))), data.frame(
  row.names=c("SIMPLE_TYPE"),
  Number=c("1"),
  Type=c("String"),
  Description=c("Simple event type annotation based purely on breakend position and orientation."))), "DataFrame"))

F_gr5 <- breakpointRanges(F_vcf5)
F_svtype5 <- simpleEventType(F_gr5)
info(F_vcf5)$SIMPLE_TYPE <- NA_character_
info(F_vcf5[F_gr5$sourceId])$SIMPLE_TYPE <- F_svtype5
info(F_vcf5[F_gr5$sourceId])$SVLEN <- F_gr5$svLen
writeVcf(F_vcf5, "GRIDSS_females_B05.sv.annotated.vcf")

F_gr5 <- F_gr5[F_gr5$FILTER == "PASS" & partner(F_gr5)$FILTER == "PASS"]
simple_F_gr5 <- F_gr5[simpleEventType(F_gr5) %in% c("INS", "INV", "DEL", "DUP")]
F_simplebed5 <- data.frame(
  chrom=seqnames(simple_F_gr5),
  start=as.integer((start(simple_F_gr5) + end(simple_F_gr5)) / 2),
  end=as.integer((start(partner(simple_F_gr5)) + end(partner(simple_F_gr5))) / 2),
  name=simpleEventType(simple_F_gr5),
  score=simple_F_gr5$QUAL,
  strand="."
)

F_simplebed5 <- F_simplebed5[F_simplebed5$start < F_simplebed5$end,]
write.table(F_simplebed5, "GRIDSS_females_B05.simple.bed", quote=FALSE, sep='\t', row.names=FALSE, col.names=FALSE)

#####################################################################################################################

F_vcf6 <- readVcf("vcf_female/Females_B06.vcf.gz", "/media/jana/Data/kakpo125/genome/kakapo_full_ref.fa")
info(header(F_vcf6)) = unique(as(rbind(as.data.frame(info(header(F_vcf6))), data.frame(
  row.names=c("SIMPLE_TYPE"),
  Number=c("1"),
  Type=c("String"),
  Description=c("Simple event type annotation based purely on breakend position and orientation."))), "DataFrame"))

F_gr6 <- breakpointRanges(F_vcf6)
F_svtype6 <- simpleEventType(F_gr6)
info(F_vcf6)$SIMPLE_TYPE <- NA_character_
info(F_vcf6[F_gr6$sourceId])$SIMPLE_TYPE <- F_svtype6
info(F_vcf6[F_gr6$sourceId])$SVLEN <- F_gr6$svLen
writeVcf(F_vcf6, "GRIDSS_females_B06.sv.annotated.vcf")

F_gr6 <- F_gr6[M_gr6$FILTER == "PASS" & partner(F_gr6)$FILTER == "PASS"]
simple_F_gr6 <- F_gr6[simpleEventType(F_gr6) %in% c("INS", "INV", "DEL", "DUP")]
F_simplebed6 <- data.frame(
  chrom=seqnames(simple_F_gr6),
  start=as.integer((start(simple_F_gr6) + end(simple_F_gr6)) / 2),
  end=as.integer((start(partner(simple_F_gr6)) + end(partner(simple_F_gr6))) / 2),
  name=simpleEventType(simple_F_gr6),
  score=simple_F_gr6$QUAL,
  strand="."
)

F_simplebed6 <- F_simplebed6[F_simplebed6$start < F_simplebed6$end,]
write.table(F_simplebed6, "GRIDSS_females_B06.simple.bed", quote=FALSE, sep='\t', row.names=FALSE, col.names=FALSE)

#####################################################################################################################

F_vcf7 <- readVcf("vcf_female/Females_B07.vcf.gz", "/media/jana/Data/kakpo125/genome/kakapo_full_ref.fa")
info(header(F_vcf7)) = unique(as(rbind(as.data.frame(info(header(F_vcf7))), data.frame(
  row.names=c("SIMPLE_TYPE"),
  Number=c("1"),
  Type=c("String"),
  Description=c("Simple event type annotation based purely on breakend position and orientation."))), "DataFrame"))

F_gr7 <- breakpointRanges(F_vcf7)
F_svtype7 <- simpleEventType(F_gr7)
info(F_vcf7)$SIMPLE_TYPE <- NA_character_
info(F_vcf7[M_gr7$sourceId])$SIMPLE_TYPE <- F_svtype7
info(F_vcf7[M_gr7$sourceId])$SVLEN <- F_gr7$svLen
writeVcf(F_vcf7, "GRIDSS_females_B07.sv.annotated.vcf")

F_gr7 <- F_gr7[F_gr7$FILTER == "PASS" & partner(F_gr7)$FILTER == "PASS"]
simple_F_gr7 <- F_gr7[simpleEventType(F_gr7) %in% c("INS", "INV", "DEL", "DUP")]
F_simplebed7 <- data.frame(
  chrom=seqnames(simple_F_gr7),
  start=as.integer((start(simple_F_gr7) + end(simple_F_gr7)) / 2),
  end=as.integer((start(partner(simple_F_gr7)) + end(partner(simple_F_gr7))) / 2),
  name=simpleEventType(simple_F_gr7),
  score=simple_F_gr7$QUAL,
  strand="."
)

F_simplebed7 <- F_simplebed7[F_simplebed7$start < F_simplebed7$end,]
write.table(F_simplebed7, "GRIDSS_females_B07.simple.bed", quote=FALSE, sep='\t', row.names=FALSE, col.names=FALSE)
